### PR TITLE
[ASM] Add request.tainted metric

### PR DESCRIPTION
**Pull request by NachoEchevarria**
## Summary of changes

This PR contains the implementation of a new metric "request.tainted", which records all the tainted objects at the end of a request. It records the tainted values that are contained in the tainting map, which might not be the same as the amount of tainted objects that have been tainted during the whole request (the map purges unused values from time to time and might delete some values in the case of hash collision). It adds a telemetry value inside the span of the request containing the amount of tainted elements for that request and it also sends a metric to the telemetry API every heartbeat, containing all the values tainted in all the requests for that period.

For all languages, we were sending the source tag with a "." separator instead of "_". This needs to be updated and has been updated in this PR.

A small change has been done in the common_metrics.json file because this metric was already defined in that file, but it was including a vulnerability tag for that metric, which is not correct. This change will need to be applied in the repository https://github.com/DataDog/dd-go and will affect to all the Iast teams.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4610)](https://github.com/DataDog/dd-trace-dotnet/tree/1ef3ab7e8089485dda4b844c25b2e1e7eee49a79)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/8c781ea73b0bfeb8fd98a291d420ec3fe7abde2c)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)
- [benchmarks/2.37.0](https://github.com/DataDog/dd-trace-dotnet/tree/1cbf0f91defaa3fa262540d88bee223bc4a86732)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4610) (11.492M)   : 0, 11492439
    master (11.239M)   : 0, 11239346
    benchmarks/2.37.0 (10.984M)   : 0, 10983597
    benchmarks/2.9.0 (11.237M)   : 0, 11237158

    section Automatic
    This PR (4610) (7.807M)   : 0, 7806549
    master (7.864M)   : 0, 7863833
    benchmarks/2.37.0 (7.620M)   : 0, 7620446
    benchmarks/2.9.0 (7.980M)   : 0, 7979544

    section Trace stats
    master (7.823M)   : 0, 7822764
    benchmarks/2.37.0 (7.601M)   : 0, 7601244

    section Manual
    This PR (4610) (10.092M)   : 0, 10092012
    master (10.001M)   : 0, 10000648
    benchmarks/2.37.0 (9.672M)   : 0, 9672017

    section Manual + Automatic
    This PR (4610) (7.566M)   : 0, 7566222
    master (7.548M)   : 0, 7547695
    benchmarks/2.37.0 (7.271M)   : 0, 7270501

    section Version Conflict
    master (6.723M)   : 0, 6723060
    benchmarks/2.37.0 (6.624M)   : 0, 6624248

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4610) (9.710M)   : 0, 9709771
    master (9.679M)   : 0, 9678607
    benchmarks/2.37.0 (9.603M)   : 0, 9603393
    benchmarks/2.9.0 (9.575M)   : 0, 9574635

    section Automatic
    This PR (4610) (6.653M)   : 0, 6653367
    master (6.864M)   : 0, 6864284
    benchmarks/2.37.0 (6.484M)   : 0, 6483840

    section Trace stats
    master (6.863M)   : 0, 6863294
    benchmarks/2.37.0 (6.737M)   : 0, 6736780

    section Manual
    This PR (4610) (8.366M)   : 0, 8366018
    master (8.743M)   : 0, 8743073
    benchmarks/2.37.0 (8.509M)   : 0, 8508994

    section Manual + Automatic
    This PR (4610) (6.438M)   : 0, 6437569
    master (6.636M)   : 0, 6636247
    benchmarks/2.37.0 (6.542M)   : 0, 6542042

    section Version Conflict
    master (5.812M)   : 0, 5811903
    benchmarks/2.37.0 (5.769M)   : 0, 5769174

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4610) (9.356M)   : 0, 9356010
    master (9.307M)   : 0, 9306588
    benchmarks/2.37.0 (9.641M)   : 0, 9641281
    benchmarks/2.9.0 (9.346M)   : 0, 9345561

    section Automatic
    This PR (4610) (6.737M)   : 0, 6736524
    master (4.429M)   : 0, 4429469
    benchmarks/2.37.0 (6.809M)   : 0, 6809282
    benchmarks/2.9.0 (6.752M)   : 0, 6751930

    section Trace stats
    master (6.582M)   : 0, 6582183
    benchmarks/2.37.0 (6.650M)   : 0, 6649920

    section Manual
    This PR (4610) (8.322M)   : 0, 8321588
    master (8.260M)   : 0, 8259686
    benchmarks/2.37.0 (8.352M)   : 0, 8352046

    section Manual + Automatic
    This PR (4610) (6.524M)   : 0, 6524437
    master (6.462M)   : 0, 6462403
    benchmarks/2.37.0 (6.611M)   : 0, 6611182

    section Version Conflict
    master (5.973M)   : 0, 5973007
    benchmarks/2.37.0 (6.067M)   : 0, 6067045

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4610) (7.732M)   : 0, 7731794
    master (7.204M)   : 0, 7204398
    benchmarks/2.37.0 (7.626M)   : 0, 7626233
    benchmarks/2.9.0 (7.928M)   : 0, 7928406

    section No attack
    This PR (4610) (2.186M)   : 0, 2186265
    master (2.090M)   : 0, 2090301
    benchmarks/2.37.0 (2.192M)   : 0, 2192041
    benchmarks/2.9.0 (3.365M)   : 0, 3365217

    section Attack
    This PR (4610) (1.737M)   : 0, 1736580
    master (1.647M)   : 0, 1646927
    benchmarks/2.37.0 (1.697M)   : 0, 1696915
    benchmarks/2.9.0 (2.632M)   : 0, 2631585

    section Blocking
    This PR (4610) (3.212M)   : 0, 3211654
    master (3.015M)   : 0, 3014683
    benchmarks/2.37.0 (3.146M)   : 0, 3146410

```

File: tracer/test/snapshots/Iast.PathTraversal.AspNetCore2.TelemetryEnabled.verified.txt
Comment (NachoEchevarria) [inline tracer/test/snapshots/Iast.PathTraversal.AspNetCore2.TelemetryEnabled.verified.txt:63]: The amount of tainted values at the end of a request will change easily and any controller code change or even test change could lead to different values, so we just check that a tag with some value is included.

File: tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:148]: Rather than doing this `Replace()` on every request (generating a lot of garbage) it might be better to just have a separate `SourceType.GetAsTag(SourceTypeName)` method that returns the correct value.

You can always write a unit test to make sure they're funtionally equivalent to this 🙂

**Review by andrewlock — APPROVED** [review]

File: tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs
Comment (NachoEchevarria) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:148]: Thanks for the suggestion! I have updated the code and added a unit test.

File: tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs
Comment (daniel-romano-DD) [inline tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:101]: It's conter intuitive that off is not the smallest value...
`internal enum IastMetricsVerbosityLevel
{
    /// <summary>
    /// Debug
    /// </summary>
    Debug = 0,

    /// <summary>
    /// The default log level
    /// </summary>
    Information = 1,

    /// <summary>
    /// Only mandatory metrics
    /// </summary>
    Mandatory = 2,

    /// <summary>
    /// Disabled
    /// </summary>
    Off = 3,
}`

Comment (robertpi) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:101]: Agree, are the values specified in the RFC?

File: tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json
Comment (robertpi) [inline tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json:522]: what's this file for? it generated or manually maintained. I guess the tag was incorrectly applied before?

**Review by robertpi — APPROVED** [review]
You should address Dani's comment, but otherwise it's good for me.

File: tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json
Comment (NachoEchevarria) [reply tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json:522]: It is a copy of the file that is defined in https://github.com/DataDog/dd-go and defines the metric values used. It was wrong : this topic was discussed with the Iast team and they agreed in this change. After merging, I will need to create a separate PR in https://github.com/DataDog/dd-go and get an approval from the telemetry team.

File: tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs
Comment (NachoEchevarria) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:101]: It follows the order of the values as they are written in the RFC:
image

Comment (NachoEchevarria) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:101]: It does not look very intuitive, agree, but I wanted to stay aligned with the values provided in the RFC. In the java code, they do the same:

image

Comment (NachoEchevarria) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs:101]: I think that this internal enum values were defined like that to stay aligned with DirectSubmissionLogLevel, which follows this pattern. Still I think that it is true that a reversed order is more intuitive. I will update the code.

File: tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json
Comment (NachoEchevarria) [reply tracer/test/Datadog.Trace.Tests/Telemetry/Metrics/common_metrics.json:522]: This file is common with other teams (we copy it, but after that, we need to update the original file in another repository), the metric was already defined, it will need an update after merging.

**Review by daniel-romano-DD — APPROVED** [review]
LGTM

File: tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs
Comment (andrewlock) [reply tracer/src/Datadog.Trace/Iast/Telemetry/ExecutedTelemetryHelper.cs]: I disagree, I think it _is_ intuitive to be this way 😉 

This is how most logging libraries work (Microrosoft, Serilog etc)

Think of it like this - you get fewer and fewer logs the higher the enum value

**Pull request merged**
